File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        Arrays.sort(times);
        long low = times[0];
        long high = (long) n * times[times.length - 1];
        long answer = 0;
        
        while(low <= high){
            long mid = (low + high) / 2;
            long p = 0;
            for(int t : times)
                p += mid / t;
            if(p >= n) {
                answer = mid;
                high = mid - 1;
            }
            else
                low = mid + 1;
        }
        
        return answer;
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    answer = 0
    times.sort()
    s, e = times[0], n*times[0]
    while s <= e:
        m = (s + e) // 2
        p = 0
        for time in times:
            p += m // time
        
        if p >= n:
            answer = m
            e = m - 1
        else:
            s = m + 1
    
    return answer